dbms/src/IO/WriteBufferFromS3.cpp

```diff
@@ -0,0 +1,285 @@
+#include <IO/WriteBufferFromS3.h>
+
+#include <IO/WriteHelpers.h>
+
+#include <Poco/DOM/AutoPtr.h>
+#include <Poco/DOM/DOMParser.h>
+#include <Poco/DOM/Document.h>
+#include <Poco/DOM/NodeList.h>
+#include <Poco/SAX/InputSource.h>
+
+#include <common/logger_useful.h>
+
+
+#define DEFAULT_S3_MAX_FOLLOW_PUT_REDIRECT 2
+#define S3_SOFT_MAX_PARTS 10000
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int INCORRECT_DATA;
+}
+
+
+WriteBufferFromS3::WriteBufferFromS3(
+    const Poco::URI & uri_,
+    size_t minimum_upload_part_size_,
+    const ConnectionTimeouts & timeouts_,
+    const Poco::Net::HTTPBasicCredentials & credentials, size_t buffer_size_
+)
+    : BufferWithOwnMemory<WriteBuffer>(buffer_size_, nullptr, 0)
+    , uri {uri_}
+    , minimum_upload_part_size {minimum_upload_part_size_}
+    , timeouts {timeouts_}
+    , auth_request {Poco::Net::HTTPRequest::HTTP_PUT, uri.getPathAndQuery(), Poco::Net::HTTPRequest::HTTP_1_1}
+    , temporary_buffer {std::make_unique<WriteBufferFromString>(buffer_string)}
+    , last_part_size {0}
+{
+    if (!credentials.getUsername().empty())
+        credentials.authenticate(auth_request);
+
+    initiate();
+}
+
+
+void WriteBufferFromS3::nextImpl()
+{
+    if (!offset())
+        return;
+
+    temporary_buffer->write(working_buffer.begin(), offset());
+
+    last_part_size += offset();
+
+    if (last_part_size > minimum_upload_part_size)
+    {
+        temporary_buffer->finish();
+        writePart(buffer_string);
+        last_part_size = 0;
+        temporary_buffer = std::make_unique<WriteBufferFromString>(buffer_string);
+    }
+}
+
+
+void WriteBufferFromS3::finalize()
+{
+    temporary_buffer->finish();
+    if (!buffer_string.empty())
+    {
+        writePart(buffer_string);
+    }
+
+    complete();
+}
+
+
+WriteBufferFromS3::~WriteBufferFromS3()
+{
+    try
+    {
+        next();
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+}
+
+
+void WriteBufferFromS3::initiate()
+{
+    // See https://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadInitiate.html
+    Poco::Net::HTTPResponse response;
+    std::unique_ptr<Poco::Net::HTTPRequest> request;
```

Comment:
Maybe `request_ptr` better? It is hard to understand that `request` is not a request, but `*request` is, with such amount of `auto`

Reply:
64d49e4565